src/stack.ts

```diff
@@ -14,130 +14,157 @@
+const STACK_SYMBOL = Symbol.for('@pulumi/cdk.Stack');
+export type create = (scope: App) => void;
+
+export class App extends AppComponent<AppConverter> implements ICloudAssemblyDirectoryProducer {
     /** @internal */
     name: string;
 
-    /** @internal */
-    converter: AppConverter;
+    public assemblyDir!: string;
 
-    /** @internal */
-    assemblyDir: string;
+    // /** @internal */
+    converter: Promise<AppConverter>;
 
-    options?: StackOptions;
+    private _app?: cdk.App;
 
-    constructor(public readonly stack: Stack) {
-        super('cdk:index:Stack', stack.node.id, {}, stack.options);
-        this.options = stack.options;
+    public appOptions?: AppOptions;
 
-        this.name = stack.node.id;
+    public get app(): cdk.App {
+        if (!this._app) {
+            throw new Error('cdk.App has not been created yet');
+        }
+        return this._app!;
+    }
 
-        const assembly = stack.app.synth();
-        this.assemblyDir = assembly.directory;
-        debug(`ASSEMBLY_DIR: ${this.assemblyDir}`);
+    /**
+     * The collection of outputs from the AWS CDK Stack represented as Pulumi Outputs.
+     * Each CfnOutput defined in the AWS CDK Stack will populate a value in the outputs.
+     */
+    public outputs: Promise<{ [outputId: string]: pulumi.Output<any> }>;
+
+    private appProps?: cdk.AppProps;
+
+    constructor(id: string, private readonly createFunc: create, props?: AppOptions) {
+        super(id, props);
+        this.appOptions = props;
+
+        this.name = id;
+        this.appProps = props?.props;
+        this.converter = this.getData();
+
+        this.outputs = this.converter.then((converter) => {
+            const stacks = Array.from(converter.stacks.values());
+            return stacks.reduce((prev, curr) => {
+                const outputs: { [outputId: string]: pulumi.Output<any> } = {};
+                for (const [outputId, args] of Object.entries(curr.stack.outputs ?? {})) {
+                    outputs[outputId] = curr.processIntrinsics(args.Value);
+                }
+                return {
+                    ...prev,
+                    ...outputs,
+                };
+            }, {});
+        });
+        this.registerOutputs(this.outputs);
+    }
 
-        debug(JSON.stringify(debugAssembly(assembly)));
+    protected async initialize(): Promise<AppConverter> {
+        const cli = AwsCdkCli.fromCloudAssemblyDirectoryProducer(this);
+        await cli.synth({ quiet: true });
 
-        this.converter = new AppConverter(this);
-        this.converter.convert();
+        const converter = new AppConverter(this);
+        converter.convert();
 
-        this.registerOutputs(stack.outputs);
+        return converter;
     }
 
-    /** @internal */
-    registerOutput(outputId: string, output: any) {
-        this.stack.outputs[outputId] = pulumi.output(output);
+    async produce(context: Record<string, any>): Promise<string> {
```

Comment:
This is what will get executed multiple times.

Reply:
This sounds like piece-de-resistane in this PR. So it's calling app.synth() and returning dir.

So couple of low-level questions here perhaps.

1. if this method fails, are we OK? do we expect it to fail or we don't expect it to regularly fail?
2. if app.synth() creates temp dirs with assets, what cleans it up? is it something to worry about?
3. the state of this is modified `this.assemblyDir = dir` and `this.stacks[child.artifactId] = child` - do we need to worry about something concurrently accessing the partial state before the framework is done calling us repeatedly or it all works out as needed?

Reply:
> if this method fails, are we OK? do we expect it to fail or we don't expect it to regularly fail?

This is called when we call `cli.synth()` as part of `initialize` and we do catch errors there.

> if app.synth() creates temp dirs with assets, what cleans it up? is it something to worry about?

Maybe? It will create it as a system tmp dir which gets cleaned up whenever the computer restarts. 

> the state of this is modified this.assemblyDir = dir and this.stacks[child.artifactId] = child - do we need to worry about something concurrently accessing the partial state before the framework is done calling us repeatedly or it all works out as needed?

We shouldn't need to worry about it. This is called in blocking code as part of initialize.

Reply:
System temp dir should be fine, yes. All good then.